packages/solana/lib/src/subscription_client/subscription_client.dart

```diff
@@ -276,27 +156,121 @@ class SubscriptionClient {
   /// validator.
   ///
   /// Returns a [Stream] that can be used to listen for events. By cancelling the
-  /// subscription returned by the `Stream.listen()` method of this stream you
-  /// can send the _unsubscribe_ JSON RPC message to cancel the subscription cleanly
+  /// subscription returned by the `Stream.listen()` method of this stream
+  /// client will automatically send `slotUnsubscribe` message;
   /// from Solana.
   ///
-  /// For [commitment] parameter description [see this document][see this document]
+  /// For [commitment] parameter description [see this document][1].
+  ///
   /// [Commitment.processed] is not supported as [commitment].
   ///
-  /// [see this document]: https://docs.solana.com/developing/clients/jsonrpc-api#configuring-state-commitment
+  /// [1]: https://docs.solana.com/developing/clients/jsonrpc-api#configuring-state-commitment
   Stream<Slot> slotSubscribe() => _subscribe<Slot>('slot');
 
   /// Subscribe to receive notification anytime a new root is set by the
   /// validator.
   ///
   /// Returns a [Stream] that can be used to listen for events. By cancelling the
-  /// subscription returned by the `Stream.listen()` method of this stream you
-  /// can send the _unsubscribe_ JSON RPC message to cancel the subscription cleanly
+  /// subscription returned by the `Stream.listen()` method of this stream
+  /// client will automatically send `rootUnsubscribe` message;
   /// from Solana.
   ///
-  /// For [commitment] parameter description [see this document][see this document]
+  /// For [commitment] parameter description [see this document][1].
+  ///
   /// [Commitment.processed] is not supported as [commitment].
   ///
-  /// [see this document]: https://docs.solana.com/developing/clients/jsonrpc-api#configuring-state-commitment
+  /// [1]: https://docs.solana.com/developing/clients/jsonrpc-api#configuring-state-commitment
   Stream<int> rootSubscribe() => _subscribe<int>('root');
+
+  /// Dispose this object and cancel any existing subscription.
+  void close() {
+    _sink.close();
+  }
+
+  static int _requestId = 1;
+
+  late final WebSocketSink _sink;
+  late final Stream<dynamic> _stream;
+
+  Stream<T> _subscribe<T>(
+    String method, {
+    List<dynamic>? params,
+    bool singleShot = false,
+  }) {
+    final controller = StreamController<T>();
+
+    controller.onListen = () {
+      int? subscriptionId;
+      final requestId = _requestId++;
+
+      late final StreamSubscription subscription;
+
+      subscription = _stream.listen(
+        (event) {
```

Comment:
Could we make this a method? I think it would be cleaner.

Reply:
I don't think it will make it cleaner – it uses scoped variables (`requestId` and `subscriptionId`).